services/knack_street_seg_updater.py

```diff
@@ -0,0 +1,212 @@
+            obj=config["object"],
+            method="update",
+            data=record,
+        )
+    except Exception as e:
+        raise Exception(e.response.text)
+
+def main(args):
+    app_name = args.app_name
+    container = args.container
+    logger.info(args)
+    config = CONFIG[app_name][container]
+
+    # Get token for AGOL
+    token = create_login_token()
+
+    # Get knack records on or after the provided modified date/time
+    filters = utils.knack.date_filter_on_or_after(
+        args.date, config['modified_date_field'], tzinfo=APP_TIMEZONE, use_time=True
+    )
+    app = knackpy.App(app_id=APP_ID, api_key=API_KEY)
+    records = app.get(container, filters=filters)
+    records_formatted = [record.format() for record in records]
+    field_mapping = create_knack_field_mapping(records[0])
+
+    if not records_formatted:
+        logger.info('No records to update. Doing nothing.')
+        return 0
+
+    unmatched_segments = []
+    for street_segment in records_formatted:
+        features = query_atx_street(
+            street_segment[config["primary_key"]], token
+        )
+        time.sleep(1) # wait one second between requests to prevent being rate limited
+        if features.get("error"):
+            # AGOL returns code 200 even for error queries.
+            raise Exception(str(features))
+
+        # handling returned segment features from AGOL
+        if features.get("features"):
+            if len(features["features"]) > 0:
+                segment_data = features["features"][0]["attributes"]
+                #  we don't want to compare modified dates
+                #  because we don't keep that value in sync with the source data on AGOL
+                #  because we use our own modified date set in the data tracker
+                segment_data.pop(config["modified_date_col_name"])
+                street_segment.pop(config["modified_date_col_name"])
+
+                #  compare new data (segment data) against old (street_segment)
+                #  we only want to upload values that have changed
+                if not are_equal(street_segment, segment_data):
+                    logger.info(f'Change detected for segment ID: {street_segment[config["primary_key"]]}')
+                    segment_data["id"] = street_segment["id"]
+                    segment_data[config["modified_date_col_name"]] = local_timestamp()
+                    # Uploading updated data back to Knack
+                    update_record(config, segment_data, field_mapping)
+            else:
+                unmatched_segments.append(street_segment[config["primary_key"]])
+                continue
+        else:
+            unmatched_segments.append(street_segment[config["primary_key"]])
+            continue
+
+    if unmatched_segments:
+        error_text = "Unmatched street segments: {}".format(
+            ", ".join(str(x) for x in unmatched_segments)
+        )
+        raise Exception(error_text)
```

Comment:
This is currently raising this error every time because there are a few old/malformed segments in data tracker today.

Reply:
do we need to open an issue to fix the old/malformed segments? or is the apps team already aware?